### PR TITLE
Currencies Module

### LINKED ISSUE #4: Tokens Module
## Goal

Create a `orml-tokens` runtime module to support multiple types of token.

### Version 1 - Basic tokens functionalities

-  [x] Minimal `trait` abstraction `MultiCurrency`.
  - `transfer`
  - `mint`
  - `burn`
  - `slash`
- [x] Runtime module.
  - Impl `MultiCurrency`.
  - Dispatchable `transfer` extrinsic
- [x] Unit test.

### Version 2 - Compatibility with `srml-balance` module

- [x] Be compatible with `srml-balance` and `Currency`, which would be used for native token.

~~### Optional - Better issuance control via "imbalance"~~

~~- [ ] "imbalance" types represents `balance` change of tokens.~~
~~- [ ] Add imbalance types and rebalance feature to `MultiCurrency` and `orml-tokens`.~~

### PR DESCRIPTION
**Pull request by shaopengw**
impl currencies module in #8, closes #4

### CODE REVIEW

File: currencies/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		Transferred(CurrencyId, AccountId, AccountId, Balance),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		/// Transfer some balance to another account.
+		pub fn transfer(
```
Comment (xlc) [inline currencies/src/lib.rs]: should have transfer handle both native and non native currency and use `T:: NativeCurrency ` for native transfer

File: currencies/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+type BalanceOf<T> = <<T as Trait>::MultiCurrency as MultiCurrency<<T as srml_system::Trait>::AccountId>>::Balance;
+type CurrencyIdOf<T> = <<T as Trait>::MultiCurrency as MultiCurrency<<T as srml_system::Trait>::AccountId>>::CurrencyId;
+type ErrorOf<T> = <<T as Trait>::MultiCurrency as MultiCurrency<<T as srml_system::Trait>::AccountId>>::Error;
+
+type AmountOf<T> = <<T as Trait>::MultiCurrency as MultiCurrencyExtended<<T as srml_system::Trait>::AccountId>>::Amount;
+
+pub trait Trait: system::Trait {
+	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+	type MultiCurrency: MultiCurrencyExtended<Self::AccountId>;
+	type GetNativeCurrencyId: Get<CurrencyIdOf<Self>>;
```
Comment (xlc) [inline currencies/src/lib.rs]: `type NativeCurrency: BasicCurrency`

File: currencies/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	}
+}
+
+impl<T: Trait> MultiCurrency<T::AccountId> for Module<T> {
+	type Balance = BalanceOf<T>;
+	type CurrencyId = CurrencyIdOf<T>;
+	type Error = ErrorOf<T>;
+
+	fn total_issuance(currency_id: Self::CurrencyId) -> Self::Balance {
+		T::MultiCurrency::total_issuance(currency_id)
```
Comment (xlc) [inline currencies/src/lib.rs]: all those methods should check if currency_id == native and delegate to `T::NativeCurrency` instead

**Comment by xlc** [comment]
Please merge master for CI fix.

**Comment by shaopengw** [comment]
The `BasicCurrency` impl for `srml_support::traits::Currency` has to be in `traits` module because of the orphan rule.

With `impl<AccountId, T: SrmlCurrency<AccountId>> BasicCurrency<AccountId> for T { ... }`, implementing `BasicCurrency` for a generic `T: srml_support::traits::Currency` would conflict with the `BasicCurrency` impl for [`Currency` struct](https://github.com/laminar-protocol/open-runtime-module-library/blob/0e13df4278e537b52c37f7fdd0a83241b25cf51b/currencies/src/lib.rs#L172), because of the restriction [E0119](https://doc.rust-lang.org/error-index.html#E0119).

A possible solution is adding another trait bound for the generic `T`, as long as `Currency` struct doesn't impl it, there would be no conflicts. `dispatch::ModuleErrorMetadata` looks like good choice, since runtime modules implement it and there is no additional generic type.

**Comment by xlc** [comment]
I am actually thinking about something like this

```
struct BasicCurrencyAdpater<C>;

impl<AccountId, C> BasicCurrency for BasicCurrencyAdpater<C>
where C: Currency<AccountId>
```

and to use

```
impl auction::Triat for Runtime {
    type Currency = BasicCurrencyAdpater<Balances>;
}
```

**Comment by shaopengw** [comment]
@xlc Updated to adapter implementation.

File: currencies/src/lib.rs [hunk 4]
```diff
@@ -218,4 +223,51 @@ where
...
+	fn balance(who: &AccountId) -> Self::Balance {
+		T::total_balance(who)
+	}
+
+	fn transfer(from: &AccountId, to: &AccountId, amount: Self::Balance) -> result::Result<(), Self::Error> {
+		T::transfer(from, to, amount, ExistenceRequirement::AllowDeath)
+	}
+
+	fn deposit(who: &AccountId, amount: Self::Balance) -> result::Result<(), Self::Error> {
+		let imbalance = T::deposit_creating(who, amount);
```
Comment (xlc) [inline currencies/src/lib.rs]: `let _ = T::deposit_creating(who, amount);` should be the same

File: currencies/src/lib.rs [hunk 5]
```diff
@@ -218,4 +223,51 @@ where
...
+	}
+
+	fn withdraw(who: &AccountId, amount: Self::Balance) -> result::Result<(), Self::Error> {
+		T::withdraw(
+			who,
+			amount,
+			WithdrawReason::Transfer.into(),
+			ExistenceRequirement::AllowDeath,
+		)
+		.map(|imbalance| {
```
Comment (xlc) [inline currencies/src/lib.rs]: I don't think you have to call drop, so just `.map(|_| ())` should be good enough

File: currencies/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+			who,
+			amount,
+			WithdrawReason::Transfer.into(),
+			ExistenceRequirement::AllowDeath,
+		)
+		.map(|_| ())
+	}
+
+	fn slash(who: &AccountId, amount: Self::Balance) -> Self::Balance {
+		let (imbalance, gap) = T::slash(who, amount);
```
Comment (xlc) [inline currencies/src/lib.rs]:
```suggestion
		let (_, gap) = T::slash(who, amount);
```

**Pull request merged**